lib/motions/general-motions.coffee

```diff
@@ -451,54 +451,50 @@ class MoveToMiddleOfScreen extends MoveToScreenLine
     height = lastScreenRow - firstScreenRow
     Math.floor(firstScreenRow + (height / 2))
 
-class ScrollKeepingCursor extends MoveToLine
-  previousFirstScreenRow: 0
-  currentFirstScreenRow: 0
+class ScrollKeepingCursor extends Motion
+  operatesLinewise: true
+  cursorRow: null
 
   constructor: (@editorElement, @vimState) ->
     super(@editorElement.getModel(), @vimState)
 
   select: (count, options) ->
-    finalDestination = @scrollScreen(count)
+    scrollTop = @scrollScreen(count)
     super(count, options)
-    @editorElement.setScrollTop(finalDestination)
+    @editorElement.setScrollTop(scrollTop)
 
   execute: (count) ->
-    finalDestination = @scrollScreen(count)
+    scrollTop = @scrollScreen(count)
     super(count)
-    @editorElement.setScrollTop(finalDestination)
+    @editorElement.setScrollTop(scrollTop)
 
-  moveCursor: (cursor, count=1) ->
-    cursor.setScreenPosition([@getDestinationRow(count), 0])
-
-  getDestinationRow: (count) ->
-    {row, column} = @editor.getCursorScreenPosition()
-    @currentFirstScreenRow - @previousFirstScreenRow + row
+  moveCursor: (cursor) ->
+    cursor.setScreenPosition(Point(@cursorRow, 0), autoscroll: false)
 
   scrollScreen: (count=1) ->
-    @previousFirstScreenRow = @editorElement.getFirstVisibleScreenRow()
-    destination = @scrollDestination(count)
-    @editorElement.setScrollTop(destination)
-    @currentFirstScreenRow = @editorElement.getFirstVisibleScreenRow()
-    destination
+    # TODO: We need to create an API for reading an editor's logical scroll
+    # position which may not yet have been flushed to the DOM. Currently, the
+    # only way to access this information is through this piece of private state.
+    currentScrollTop = @editorElement.component.presenter.pendingScrollTop ? @editorElement.getScrollTop()
```

Comment:
Thanks @maxbrunsfeld for bringing this up. So, there are a couple of ways in which we could solve this. We may either:

1. Add something like a `TextEditorElement::getPendingScrollTop` API;
2. Flush pending state (such as scroll positions) synchronously whenever methods like `::getScrollTop` get called;

Personally, I would prefer 2) over 1) because it both feels more consistent with how the DOM works like and it doesn't expose an internal information which heavily depends on the update scheduling implementation. For example: what does pending mean here? In the current implementation we know that the scroll position's state is either pending or committed, but what if we started handling this differently by introducing a third state?

On the other hand, I understand the implications of having such a synchronous API: forcing reflows is not ideal, especially when the interaction with the user needs to happen repeatedly such as in this case (e.g. holding down a key). It's also true that people are already used to how the DOM behaves when reading/writing from/to it, so I feel like this wouldn't be a deal breaker. Indeed, this piece of code could be changed to make use of `requestAnimationFrame` (or our `ViewRegistry::readDocument` + `ViewRegistry::updateDocument` methods) to orchestrate DOM updates in a way that doesn't affect performance because of synchronous reflows.

What do you think?